### PR TITLE
Describe Engineering Best Practices

### PR DESCRIPTION
**Pull request by coverbeck**
**Description**
The issue title is "Document CI/CD best practices for workflows"; I broadened it to general software engineering practices.

It's hard to strike the right balance; I can't replace a book. I was also running a bit out of steam. Hopefully this works, although it might be a little thin.

I don't really talk about CI/CD; I feel they're technically not the same thing as automated builds, although automated builds are part of CI. But I'm open to suggestions.

**Issue**
dockstore/dockstore#3669

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [ ] If the PR has any new pages, only **AFTER** the PR is approved by all reviewers, generate a new discourse topic, by running the action at https://github.com/dockstore/dockstore-documentation/actions/workflows/add-discourse-topic.yml. Otherwise we may end up with orphaned topics, which would need manual management.

### CODE REVIEW

File: docs/advanced-topics/best-practices/best-practices-development.rst [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+`````````````
+
+There are lots of options for doing automated builds. You can search the web for "Continuous Integration" and get lots of results. As a starting point, here are some services the Dockstore team has used:
+
+- `GitHub Actions <https://docs.github.com/en/actions>`__
+- `CircleCI <https://circle.com>`__
+- `Travis CI <https://www.travis-ci.com>`__
+- `Jenkins <https://www.jenkins.io>`__
+
+GitHub Actions, CircleCI and Travis CI are cloud services; Jenkins is a server you set up in your own infrastructure.
```
Comment (denis-yuen) [inline docs/advanced-topics/best-practices/best-practices-development.rst:37]: Maybe throw some "generally" in here. I think there are companies that sell Jenkins as a service and GitHub Actions allows you to farm out jobs to your own infrastructure, but those are exceptions that "generally" probably covers our butt enough

**Review by denis-yuen — APPROVED** [review]
Nice!
I think this will be worth highlighting in our blog/feattured news

File: docs/advanced-topics/best-practices/best-practices-development.rst [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+Workflow languages do not typically make it easy to share dependencies like other languages do (Maven in Java, PyPI in Python, etc.). One common techinque is to have one Git repository with multiple workflows, with all the workflows referencing common code in the same repository, using relative paths. You can then publish multiple workflows from the single repository, or push the individual workflows to different repositories. For an example of the latter, see the Galaxy IWC use case below.
+
+Another technique is to `import code using https urls <https://github.com/aofarrel/myco/blob/469620a1c8ecda44ae843985f6d640e9ca24d028/myco_sra.wdl#L3>`__. This can work, but does get difficult to manage as you'll need to update multiple repositories if a change is made in common code.
+
+
+Use Cases
+---------
+
+Following are a couple of real world examples of workflows published on Dockstore, that use some of the techinques described above.
```
Comment (svonworl) [inline docs/advanced-topics/best-practices/best-practices-development.rst:78]: "techinques"

**Review by svonworl — COMMENTED** [review]
The content is good.  There's some room to tighten up the phrasing, use active voice, etc.  I wonder if there's a tool that analyzes a target sentence/paragraph/document and spits out a version that reads better, this seems like a task that the AI would excel at...

File: docs/advanced-topics/best-practices/best-practices-development.rst [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+Flattening and Packing
+``````````````````````
+
+Some languages offer tools that package your workflows into a single file. For example, `cwltool <https://github.com/common-workflow-language/cwltool>`__ offers a `--pack` option to combine all of a CWL workflow's files into one single file. This makes it easier to distribute your workflow, and enables your workflow to run ona platforms that do not support imports.
+
+Sharing Code
+------------
+
+Workflow languages do not typically make it easy to share dependencies like other languages do (Maven in Java, PyPI in Python, etc.). One common techinque is to have one Git repository with multiple workflows, with all the workflows referencing common code in the same repository, using relative paths. You can then publish multiple workflows from the single repository, or push the individual workflows to different repositories. For an example of the latter, see the Galaxy IWC use case below.
```
Comment (kathy-t) [inline docs/advanced-topics/best-practices/best-practices-development.rst:70]: techinque -> technique

**Review by kathy-t — APPROVED** [review]

**Comment by coverbeck** [comment]
Tried to address some of the passive voice stuff, but I think I use it so frequently I may not have caught it all. We have a ticket for a broader improvement/tooling, but I'd consider that beyond the scope of this release.

**Review by svonworl — APPROVED** [review]

**Pull request merged**